docs/design/2022-11-17-ttl-table.md

```diff
@@ -0,0 +1,412 @@
+
+- `ttl_queries`
+  - The total count of select queries in TTL jobs
+  - Type: Counter
+  - Labels: type, table, result
+
+- `ttl_processed_expired_rows`
+  - The total count of expired rows processed in TTL jobs
+  - Type: Counter
+  - Labels: type, table
+
+- `ttl_query_duration`
+  - The duration of the queries in TTL jobs
+  - Labels: type
+  - Type: Histogram
+
+- `ttl_job_status`
+  - The status for the current TTL job. When the job is in the specified status, the value will 1, otherwise, it will be 0
+  - Labels: table, status
+  - Type: Gauge
+
+- `ttl_job_scan_workers`
+  - The running scan workers count for the ttl jobs
+  - Labels: table
+  - Type: Gauge
+
+In the above metrics, the optional values for type label is 'select' and 'delete' and the optional values for result label is `success` and `error`
+
+## Known Issues
+
+- Though the TTL jobs from different table runs distributively. However, one job from a table runs in a single TiDB node. If a table is very large, there may be a bottleneck.
+- Currently, the condition of generated column cannot be pushed down to TiKV. If a table uses generated column as a TTL time column, the filter will be performed in TiDB side. It brings some necessary network traffics and makes the query slow.
+
+## Future Works
+
+### Performance Optimizations
+
+We'll do some performance optimizations in the future to reduce the execution time and performance cost of the TTL job:
+
+- If an index with the prefix of the TTL time column exists, we can use it to query expire rows instead of scanning the full table. It will reduce the execution time of the scan tasks.
+- In most times, secondary indexes will not be created to avoid the hotspot of the insert. In this scene, we can also reduce some unnecessary scans by caching the statistical information. For example, we can cache the created time of the oldest row for each region after a job finished. When the next job starts, it can check the all the regions, if the data of one region do not have any updates and its cached time is not expired, just skip that region.
+- If a TTL table has some Tiflash replicas, we can the TiFlash instead of TiKV. 
+- In the future, we can schedule the tasks from one job to multiple nodes instead of executing them only in one node. This approach will improve the resource utilization of the cluster. It also means we can execute more tasks in concurrency at the same time that makes the scan and delete faster.
+- If a table does not have any secondary index, we can do some further optimizations. One optimization is that to push down the scan and delete to TiKV side without data exchanging between TiDB and TiKV. It is somewhat like what GCWorker dose. In this way, a new coprocessor command "TTLGC" will be introduced and when a job starts, TiDB will send "TTLGC" commands to each region and TiKV will then scan and delete the expired rows (TiKV should delete expired rows in a non-transactional way).
```

Comment:
scan with stale read to avoid lock contention with main traffic write

Reply:
integrate with resource control framework?

Reply:
scan and delete with LOW_PRIORITY has already been implemented in the first version